### PR TITLE
DRMAAV2-21: Initial stubs for SessionManager

### PR DESCRIPTION
**Pull request by vinodchitrali**
#### Issue-ID

* **[DRMAAV2-21](https://pbspro.atlassian.net/browse/DRMAAV2-21)**

#### Problem description
* Initial stubs for SessionManager 

#### Cause / Analysis
* N/A 

#### Solution description
* Make SessionManager as singleton and provide stubs for all implementation

#### Checklist:

- [x] I have joined the **[pbspro community forum](http://community.pbspro.org/)**.
- [x] My pull request contains a **single, signed** commit. See **[setting up gpg signature](https://pbspro.atlassian.net/wiki/display/DG/Signing+Your+Git+Commits)**.
- [x] My code follows the **[coding style](https://pbspro.atlassian.net/wiki/display/DG/Coding+Standards)** of this project.
- [ ] My change requires project documentation. See **[required documentation checklist](https://pbspro.atlassian.net/wiki/display/DG/Checklist+for+Developing+Features+and+Bug+Fixes)** for details.
- [ ] I have added documentation in the **[project documentation area](https://pbspro.atlassian.net/wiki/display/PD)**.
- [ ] I have added new **unit test(s) to my commit**. *(or)*
- [ ] I have added  **manual test(s) to the Jira ticket and explained why unit testing is not appropriate** for this case.
- [ ] All new and existing automated tests have passed. 
- [x] I have attached **test logs to the Jira ticket** as evidence of testing/verification.

__***For further information please visit the [Developer Guide Home](https://pbspro.atlassian.net/wiki/display/DG/Developer+Guide+Home).***__

### CODE REVIEW

File: src/Makefile.am [hunk 1]
```diff
@@ -37,9 +37,9 @@
 
 noinst_LTLIBRARIES=     libsrc.la
 
-libsrc_la_SOURCES=      Session.cpp 
+libsrc_la_SOURCES=      Session.cpp  SessionManagerImpl.cpp
 
-libsrc_la_CPPFLAGS =    -I$(top_srcdir)/inc
+libsrc_la_CPPFLAGS =    -I$(top_srcdir)/inc -I$(top_srcdir)/api/cpp-binding/
```
Comment (mike0042) [inline src/Makefile.am]: For the sake of consistency, please remove the trailing /

File: test/unittesting/src/Makefile.am [hunk 2]
```diff
@@ -36,7 +36,7 @@
 #
 noinst_PROGRAMS = test_drmaa
 
-test_drmaa_SOURCES=	testdrmaa.cpp
+test_drmaa_SOURCES=	SessionManagerTest.cpp testdrmaa.cpp
```
Comment (mike0042) [inline test/unittesting/src/Makefile.am]: Please remove the trailing space.

File: api/cpp-binding/drmaa2.hpp [hunk 3]
```diff
@@ -1248,6 +1283,7 @@ class SessionManager {
 	 */
 	virtual void registerEventNotification(const DrmaaCallback& callback) = 0;
 };
-
 }
+;
```
Comment (vinodchitrali) [inline api/cpp-binding/drmaa2.hpp]: ; at the end of namespace is allowed and is optional. Will remove to maintain consistency

File: api/cpp-binding/drmaa2.hpp [hunk 4]
```diff
@@ -968,9 +971,51 @@ class MonitoringSession {
...
+
+template<typename IfaceT, typename ImplT>
+class Singleton {
+private:
+	static IfaceT * m_instance;
+	/**
+	 * @brief Copy constructor is private so that
+	 * 			instance cannot be created outside class
+	 */
+	Singleton(Singleton const&);
```
Comment (arungrover) [inline api/cpp-binding/drmaa2.hpp]: Does it not need a default constructor here?

File: src/SessionManagerImpl.cpp [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+ * "PBS Professional®", and "PBS Pro™" and Altair’s logos is subject to Altair's
+ * trademark licensing policies.
+ *
+ */
+
+#include <SessionManagerImpl.h>
+
+namespace drmaa2 {
+
+SessionManagerImpl::~SessionManagerImpl() {
```
Comment (arungrover) [inline src/SessionManagerImpl.cpp]: All functions are missing function header here.

Comment (vinodchitrali) [inline src/SessionManagerImpl.cpp]: function header ? do u mean doxygen for functions ?   SessionManagerImpl.h has doxygen documentation

File: configure.ac [hunk 6]
```diff
@@ -39,13 +39,21 @@
...
 $ versioning 
-DRMAA2_MAJOR_VERSION=2
-DRMAA2_MINOR_VERSION=2
-DRMAA2_MICRO_VERSION=0
+DRMAA_NAME="DRMAAV2"
+DRMAA_MAJOR_VERSION=2
+DRMAA_MINOR_VERSION=2
+DRMAA_MICRO_VERSION=0
+#TODO Should be from the installed DRMS
+DRMS_NAME="DMRS"
```
Comment (vinodchitrali) [inline configure.ac]: This is something we yet to decide. Both DRMS name and version should be fetched by contacting to DRMS(PBSPro). line 47 added TODO to fix later. Once we have back-end to contact DRMS then we redefine these variables.

Comment (arungrover) [inline configure.ac]: Well that makes sense. But, DRMS_NAME can still be defaulted to PBSPro if not version.

File: src/SessionManagerImpl.cpp (hunk 5 above)
Comment (arungrover) [inline src/SessionManagerImpl.cpp]: Sorry about the confusion. I meant doxygen headers like there is in drmaa2.cpp

Comment (vinodchitrali) [inline src/SessionManagerImpl.cpp]: Since C++ code needs class level doxygen as well. So for all C++ code the doxygen is part of header file. U can see the doxygen docs in  SessionManagerImpl.h

File: configure.ac (hunk 6 above)
Comment (vinodchitrali) [inline configure.ac]: @ top level(API) code doesn't make any reference to PBSPro.  

In any case will add  PBSPro

File: api/cpp-binding/drmaa2.hpp [hunk 7]
```diff
@@ -968,9 +971,51 @@ class MonitoringSession {
...
+	 */
+	Singleton& operator=(Singleton const&);
+
+public:
+	/**
+	 * @brief to get single Instance
+	 *
+	 * @return  IfaceT - Abstract class reference
+	 *
+	 */
```
Comment (vinodchitrali) [inline api/cpp-binding/drmaa2.hpp]: fixed

File: api/cpp-binding/drmaa2.hpp [hunk 8]
```diff
@@ -795,8 +798,8 @@ class JobSession {
 	 */
 	JobSession(const string& contact_, const string& sessionName_,
 			const StringList& jobCategories_) :
-			contact(contact_), sessionName(sessionName_),
-					jobCategories(jobCategories_) {
+			contact(contact_), sessionName(sessionName_), jobCategories(
+					jobCategories_) {
```
Comment (vinodchitrali) [inline api/cpp-binding/drmaa2.hpp]: fixed

File: test/unittesting/src/Makefile.am (hunk 2 above)
Comment (vinodchitrali) [inline test/unittesting/src/Makefile.am]: fixed

File: src/SessionManagerImpl.cpp (hunk 5 above)
Comment (arungrover) [inline src/SessionManagerImpl.cpp]: ok I understand now and I'm okay with having it in header file. I just wanted it to be consistent with drmaa2.cpp, which it isn't right now.

File: configure.ac [hunk 9]
```diff
@@ -39,13 +39,21 @@
...
-DRMAA2_MICRO_VERSION=0
+DRMAA_NAME="DRMAAV2"
+DRMAA_MAJOR_VERSION=2
+DRMAA_MINOR_VERSION=2
+DRMAA_MICRO_VERSION=0
+#TODO Should be from the installed DRMS
+DRMS_NAME="PBSPro"
+DRMS_MAJOR_VERSION=14
+DRMS_MINOR_VERSION=2
+DRMS_MICRO_VERSION=0
```
Comment (mike0042) [inline configure.ac]: Filed https://pbspro.atlassian.net/browse/PP-713 to include PBS Pro version in header file.

File: test/unittesting/src/Makefile.am (hunk 2 above)
Comment (mike0042) [inline test/unittesting/src/Makefile.am]: Is it fixed? I still see unnecessary whitespace on lines 39 and 40.

Comment (vinodchitrali) [inline test/unittesting/src/Makefile.am]: Fixed in latest commit

File: api/cpp-binding/drmaa2.hpp [hunk 10]
```diff
@@ -968,9 +973,59 @@ class MonitoringSession {
...
+
+public:
+	/**
+	 * @brief to get single Instance
+	 *
+	 * @return  IfaceT - Abstract class reference
+	 *
+	 */
+	static IfaceT* getInstance() {
+		if (m_instance == NULL) {
```
Comment (arungrover) [inline api/cpp-binding/drmaa2.hpp]: I'm just picking cues from other pull, does this also needs to be locked in a mutex?

Comment (vinodchitrali) [inline api/cpp-binding/drmaa2.hpp]: This cannot be locked, bcz all the methods of SessionManager will be reentrant. And methods will have local mutex

**Comment by arungrover** [comment]
Looks good @vinodchitrali please squash your commits.
Thanks!

**Comment by vinodchitrali** [comment]
squashed

**Comment by arungrover** [comment]
All looks good, merged.
